drivers/timer/arcv2_timer0.c

```diff
@@ -131,17 +163,40 @@ static ALWAYS_INLINE void timer0_limit_register_set(u32_t count)
 }
 
 #if !SMP_TIMER_DRIVER
+/* This internal function calculates the amount of HW cycles that have
+ * elapsed since the last time the absolute HW cycles counter has been
+ * updated. 'cycle_count' may be updated either by the ISR, or
+ * in z_clock_set_timeout().
+ *
+ * Additionally, the function updates the 'overflow_cyc' counter, that
+ * holds the amount of elapsed HW cycles due to (possibly) multiple
+ * timer wraps (overflows).
+ *
+ * Prerequisites:
+ * - reprogramming of LIMIT must be clearing the COUNT
+ * - ISR must be clearing the 'overflow_cyc' counter.
+ * - no more than one counter-wrap has occurred between
+ *     - the timer reset or the last time the function was called
+ *     - and until the current call of the function is completed.
+ * - the function is invoked with interrupts disabled.
+ */
 static u32_t elapsed(void)
 {
-	u32_t val, ov, ctrl;
+	u32_t val, ctrl;
 
 	do {
 		val =  timer0_count_register_get();
 		ctrl = timer0_control_register_get();
 	} while (timer0_count_register_get() < val);
 
-	ov = (ctrl & _ARC_V2_TMR_CTRL_IP) ? last_load : 0;
-	return val + ov;
+	if (ctrl & _ARC_V2_TMR_CTRL_IP) {
+		overflow_cyc += last_load;
+		/* clear the IP bit of the control register */
+		timer0_control_register_set(_ARC_V2_TMR_CTRL_NH |
```

Comment:
the overhead in elapsed() must b as small as possible to avoid cycle loss in set_timeout.
For readability concern, i can merge _ARCH_V2_TMR_CTRL_NH with other macro.

Reply:
No merging of 2 defines in one makes even less sense as it may mislead a reader.
Speaking of minimization of overhead I guess we already do a lot here as well as we'll possibly have prologue/epilogue generated by the compiler which we don't see in C-sources here. So another 2 instructions and few CPU cycles of overhead shouldn't hurt I guess.